pilot/pkg/bootstrap/server.go

```diff
@@ -1099,6 +1106,124 @@ func (s *Server) initControllers(args *PilotArgs) error {
 	return nil
 }
 
+const ( // CACerts file operation states
+	create  = 0
+	write   = 1
+	remove  = 2
+	invalid = 3
+)
+
+var cacertsMap map[string]uint8
+
+func isValidCACertsFile(file string) (string, bool) {
+	if strings.Contains(file, "ca-cert.pem") {
+		return "ca-cert.pem", true
+	} else if strings.Contains(file, "ca-key.pem") {
+		return "ca-key.pem", true
+	} else if strings.Contains(file, "root-cert.pem") {
+		return "root-cert.pem", true
+	} else if strings.Contains(file, "cert-chain.pem") {
+		return "cert-chain.pem", true
+	}
+
+	return "", false
+}
+
+func isValidCACertsEvent(event fsnotify.Event) uint8 {
+	if event.Op&fsnotify.Create == fsnotify.Create {
+		return create
+	} else if event.Op&fsnotify.Write == fsnotify.Write {
+		return write
+	} else if event.Op&fsnotify.Remove == fsnotify.Remove {
+		return remove
+	}
+
+	return invalid
+}
+
+func updateCACertsMap(event fsnotify.Event) bool {
+	file, ok := isValidCACertsFile(event.Name)
+	if !ok {
+		return false
+	}
+
+	op := isValidCACertsEvent(event)
+	if op == invalid {
+		return false
+	}
+
+	cacertsMap[file] = op
+	return true
+}
+
+func handleEvent(s *Server, event fsnotify.Event) {
+	ok := updateCACertsMap(event)
+	if !ok {
+		return
+	}
+
+	if (cacertsMap["ca-cert.pem"] == create && cacertsMap["ca-key.pem"] == create &&
+		cacertsMap["root-cert.pem"] == create && cacertsMap["cert-chain.pem"] == create) ||
+		(cacertsMap["ca-cert.pem"] == write && cacertsMap["ca-key.pem"] == write &&
+			cacertsMap["root-cert.pem"] == write && cacertsMap["cert-chain.pem"] == write) {
+		log.Info("Update Istiod cacerts")
+		err := s.CA.GetCAKeyCertBundle().UpdateNewPluggedInCACerts(
```

Comment:
can't be done this way. The CA may be in use in multiple goroutines to sign certificates. Has to be done inside a wlock() I think.

Reply:
https://github.com/istio/istio/blob/master/security/pkg/pki/util/keycertbundle.go#L158 does protect writing. right?